rasa/nlu/classifiers/diet_classifier.py

```diff
@@ -1375,16 +1398,30 @@ def _mask_loss(
         )
 
         lm_mask_bool = tf.squeeze(lm_mask_bool, -1)
-        # pick elements that were masked
+
+        # Pick elements that were masked, throwing away the batch & sequence dimension
+        # and effectively switching from shape (batch_size, sequence_length, units) to
+        # (num_masked_elements, units).
         outputs = tf.boolean_mask(outputs, lm_mask_bool)
         inputs = tf.boolean_mask(inputs, lm_mask_bool)
         ids = tf.boolean_mask(seq_ids, lm_mask_bool)
 
-        outputs_embed = self._tf_layers[f"embed.{name}_lm_mask"](outputs)
-        inputs_embed = self._tf_layers[f"embed.{name}_golden_token"](inputs)
+        tokens_predicted_embed = self._tf_layers[f"embed.{name}_lm_mask"](outputs)
+        tokens_true_embed = self._tf_layers[f"embed.{name}_golden_token"](inputs)
 
+        # To limit the otherwise computationally expensive loss calculation, we
+        # constrain the label space in MLM (i.e. token space) to only those tokens that
+        # were masked in this batch. Hence the reduced list of token embeddings
+        # (tokens_true_embed) and the reduced list of labels (ids) are passed as
+        # all_labels_embed and all_labels, respectively. In the future, we could be less
+        # restrictive and construct a slightly bigger label space which could include
+        # tokens not masked in the current batch too.
         return self._tf_layers[f"loss.{name}_mask"](
-            outputs_embed, inputs_embed, ids, inputs_embed, ids
+            inputs_embed=tokens_predicted_embed,
```

Comment:
Because the `call` of `DotProductLoss` is written such that it doesn't expect any positional arguments or any argument named `inputs`. Strictly speaking, we should change that `call` definition because it will fail if the first named argument  to `call` (in this case it's `inputs_embed`) will ever be `[]`.

Reply:
@Ghostvv I think we should have a call where we discuss exactly what the issue is and which patterns we need to avoid to prevent that weird error from happening.